clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+  kernelFunc();
+}
+
+// kernel3 - expect error
+// The current function is named with a lambda (i.e., takes a lambda as a
+// template parameter. Call the builtin on the current function
+// Then it is passed to kernel
+// Test that passing the current function to the unique stable name builtin and
+// then to the kernel throws an error because the latter causes its name mangling
+// to change
+template <typename Func>
+void kernel3func(const Func &F3) {
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3); // #USN_F3
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#kernel3func_call{{in instantiation of function template specialization}}
+  // expected-note@#USN_F3{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel3>(F3); // #kernel3_call
+}
+
+void callkernel3() {
+  kernel3func([]() {}); // #kernel3func_call
+}
+
+// kernel4 - expect error
+// The current function is named with a lambda (i.e., takes a lambda as a
+// template parameter. Call the builtin on the current function
+// Then an empty lambda is passed to kernel
+// Test that passing the current function to the unique stable name builtin and
+// then passing a lambda to the kernel throws an error because the current function
+// is part of naming the kernel although it is not passed to the kernel, and its
+// mangling changes due to kernel call with the lambda
+template <typename Func>
+void kernel4func(const Func &F4) {
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4); // #USN_F4
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#kernel4func_call{{in instantiation of function template specialization}}
+  // expected-note@#USN_F4{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel4>([]() {});
+}
+
+void callkernel4() {
+  kernel4func([]() {}); // #kernel4func_call
+}
+
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct Tangerine {};
+
+template <typename Func>
+void kernel13_14func(const Func &F) {
+  kernel_single_task<class kernel13>(F);
+  kernel_single_task<class kernel14>(F);
+  // Using the same functor twice should be fine
+}
+
+// kernel13 and kernel14 - expect no errors
+// Pass the same lambda to two kernels
+// Test that passing the same lambda to two kernels does not cause an error
+// because the kernel uses do not interfere with each other or invalidate
+// the stable name in any way
+void callkernel13_14() {
+  kernel13_14func([]() {});
+}
```

Comment:
```suggestion
template <typename Func>
void kernel13_14func(const Func &F) {
  // Tests that passing the same lambda to two kernels does not cause an error
  // because the kernel uses do not interfere with each other or invalidate
  // the stable name in any way.
  kernel_single_task<class kernel13>(F);
  kernel_single_task<class kernel14>(F);
}

// kernel13 and kernel14 - expect no errors.
void callkernel13_14() {
  kernel13_14func([]() {});
}
```